plonky2/src/fri/batch_prover.rs

```
@@ -0,0 +1,456 @@
+            lde_polynomial_values,
+            challenger,
+            fri_params,
+        )
+    );
+
+    // PoW phase
+    let pow_witness = timed!(
+        timing,
+        "find proof-of-work witness",
+        fri_proof_of_work::<F, C, D>(challenger, &fri_params.config)
+    );
+
+    // Query phase
+    let query_round_proofs = batch_fri_prover_query_rounds::<F, C, D>(
+        initial_merkle_trees,
+        &trees,
+        challenger,
+        n,
+        fri_params,
+    );
+
+    FriProof {
+        commit_phase_merkle_caps: trees.iter().map(|t| t.cap.clone()).collect(),
+        query_round_proofs,
+        final_poly: final_coeffs,
+        pow_witness,
+    }
+}
+
+pub(crate) fn batch_fri_committed_trees<
+    F: RichField + Extendable<D>,
+    C: GenericConfig<D, F = F>,
+    const D: usize,
+>(
+    mut final_coeffs: PolynomialCoeffs<F::Extension>,
+    values: &[PolynomialValues<F::Extension>],
+    challenger: &mut Challenger<F, C::Hasher>,
+    fri_params: &FriParams,
+) -> FriCommitedTrees<F, C, D> {
+    let mut trees = Vec::with_capacity(fri_params.reduction_arity_bits.len());
+    let mut shift = F::MULTIPLICATIVE_GROUP_GENERATOR;
+    let mut polynomial_index = 1;
+    let mut final_values = values[0].clone();
+    for arity_bits in &fri_params.reduction_arity_bits {
+        let arity = 1 << arity_bits;
+
+        reverse_index_bits_in_place(&mut final_values.values);
+        let chunked_values = final_values.values.par_chunks(arity).map(flatten).collect();
+        let tree = MerkleTree::<F, C::Hasher>::new(chunked_values, fri_params.config.cap_height);
+
+        challenger.observe_cap(&tree.cap);
+        trees.push(tree);
+
+        let beta = challenger.get_extension_challenge::<D>();
+        // P(x) = sum_{i<r} x^i * P_i(x^r) becomes sum_{i<r} beta^i * P_i(x).
+        final_coeffs = PolynomialCoeffs::new(
+            final_coeffs
+                .coeffs
+                .par_chunks_exact(arity)
```

This one silently drops the last chunk, if it's not an exact fit.  Is that what we want?  Can it ever happen?

---

We will not make this happen when we calculate 'arity,' which is a power of two value.
We will make sure:
sum {arity_bits} + final poly degree bits = lde poly degree bits OR
sum {arity_bits} =  lde poly degree bits - cap height

---

OK.  Then there's no different between `par_chunks` and `par_chunks_exacty` for us?  Perhaps we should only use one of the two variants in that case, otherwise the reader will be confused (I was).